README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# BayesianPower

<!-- badges: start -->

<!-- badges: end -->

The goal of BayesianPower is to determine the required sample size or
the unconditional and conditional error probabilities when a hypothesis
that imposes inequality constraints between group means is compared with
its complement, another inequality constraint hypothesis or the
unconstrained hypothesis. [Klaassen, Hoijtink & Gu
(unpublished)](10.17605/OSF.IO/D9EAJ) describe four methods of weighing
unconditional or conditional error probabilities to determine the sample
size. Reversely, the same rules can be applied to determine the power of
a study.

## Installation

You can install the released version of BayesianPower from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("BayesianPower")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("fayetteklaassen/BayesianPower")
```

## Vignette

There is a vignette available on
[GitHub](https://github.com/fayetteklaassen/BayesianPower/blob/master/vignettes/bayesianpower.Rmd)
that contains a description of the functionality of the package and some
examples.


NEWS.md
# NEWS for BayesianPower

# BayesianPower 0.2.3
The computation of the indecision probability in Approach 2b has been corrected.


# BayesianPower 0.2.2
Seed option in Power function fixed. Approach 3 methodology has been updated to output the median under H2.

# BayesianPower 0.2.1
Power and sample size determination functions now also output the 10% and 90% percentile of the Bayes factor distribution.

# BayesianPower 0.2.0
New computation for the complexity. The prior is now scaled to the posterior.
The option to consider group specific standard deviations is added.

# BayesianPower 0.1.6

## Minor changes
Doi link in vignette html to full URL.

# BayesianPower 0.1.5

## Minor changes
Doi link in vignette html.

# BayesianPower 0.1.4

## Minor changes
Doi link correction.

# BayesianPower 0.1.3

## Minor changes
End description with a full stop.

# BayesianPower 0.1.2

## Minor changes
Added reference to functionality of package in Description.

# BayesianPower 0.1.1

## Minor changes
Changed examples for submission to CRAN. Long examples are wrapped in donttest, 
short examples are added to enable quick testing of functionality.


# BayesianPower 0.1.0

## Major changes
Release of package. Contains two public functions: \code{bayes_power()} and \code{bayes_sampsize()}


tests/testthat.R
library(testthat)
library(BayesianPower)

test_check("BayesianPower")


tests/testthat/test_bayespower.R
context("Errors work")
library(BayesianPower)

test_that("No invalid entries are allowed", {
  n <- 12.3
  h1 <- matrix(c(1,2,3,3,2,1), nrow = 2, byrow = TRUE)
  h2 <- "u"
  m1 <- c(1.3, 23, 1)
  m2 <- c(1, 0, 21)
  expect_error(bayes_power(n, h1, h2, m1, m2))
})


R/utils.R
# Documentation eval_const() ====
#' Evaluate a constraint matrix for a set of prior/posterior samples
#'
#' @param hyp A constraint matrix defining a hypothesis.
#' @param samples A matrix. Prior or posterior samples, the number of columns
#' corresponds to the number of groups, the number of rows the number of samples
#' @return A number between 0 and 1. The proportion of samples in which the
#' constraints are met.
eval_const <- function(hyp, samples) {
  # Errors ====
  if(length(dim(hyp)) != 2) {
    stop("Hypothesis must be a matrix")
  }
  if(length(dim(samples)) != 2) {
    stop("Samples must be in a matrix form")
  }
  if(ncol(samples) != ncol(hyp)) {
    stop("Not the right dimensions")
  }
  # Function ====
  nconst <- nrow(hyp)

  out <- mean(apply(samples, 1, function(i) {
    rowCheck <- apply(hyp, 1, function(j) {
      sum(i * j) > 0
    }
    )
    sum(rowCheck) == nconst
  }
  )
  )
  return(out)
}

# Documentation samp_dist() ====
#' Sample from prior or posterior distribution
#'
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param means A vector. The prior or posterior means for each group
#' @param sds A number or a vector. The standard deviations for each group
#' If a number is used, the same prior or posterior standard deviation is
#' used for each group.
#' @return A matrix of \code{nsamp} rows and as many columns as the
#' length of \code{means}.
samp_dist <- function(nsamp, means, sds) {
  # Function ====
  ngroup <- length(means)

  samples <- matrix(stats::rnorm(nsamp * ngroup,
                          mean = means,
                          sd = sds),
                    ncol = ngroup,
                    byrow = TRUE)
  return(samples)
}

# Documentation calc_fc() ====
#' Compute the complexity or fit for two hypotheses.
#' @param hyp A constraint matrix defining H1.
#' @param hyp2 A constraint matrix defining H2 OR a character \code{'u'}
#' or \code{'c'} specifying an unconstrained or complement hypothesis
#' @param means A vector of posterior or prior means
#' @param sds A vector or posterior or prior standard deviation
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @return A vector.
#' The proportion of posterior samples in agreement with H1 and with H2
calc_fc <- function(hyp, hyp2, means, sds, nsamp = 1000) {
  # Errors ====
  if(is.numeric(hyp2)){
    if(length(dim(hyp2)) != 2) stop("Hypothesis must be a matrix")
  } else {
    if(hyp2 != "c" && hyp2 != "u") stop("Use 'u' or 'c' for unconstrained or complement.")
  }
  if(length(dim(hyp)) != 2) {
    stop("Hypothesis must be a matrix")
  }
  # Function ====
  ngroup <- length(means)
  samples <- samp_dist(nsamp, means = means, sds = sds)
  fc1 <- eval_const(hyp, samples)
  if (fc1 == 0) fc1 <- 1/nsamp
  fc2 <- if (is.character(hyp2)) {
    if (hyp2 == 'u') 1
    if (hyp2 == 'c') 1 - fc1
  } else {
    eval_const(hyp2, samples)
  }
  if (fc2 == 0) fc2 <- 1 / nsamp
  out <- c(fc1, fc2)
  return(out)
}

# Documentation calc_bf()====
#' Compute a Bayes factor
#'
#' @param data A matrix. The dataset for which the BF must be computed
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param scale A number specifying the prior scale.
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' @return BF12, that is, the evidence for H1 relative to H2
calc_bf <- function(data, h1, h2, scale, nsamp = 1000) {
  # Function ====
  postmeans <- colMeans(data)
  postsds <- apply(data, 1, stats::sd) / sqrt(nrow(data))
  ngroup <- length(postmeans)
  priormeans <- rep(0, ngroup)
  priorsds <- postsds*scale
  comp <- calc_fc(h1, h2, priormeans, priorsds, nsamp)
  fit <- calc_fc(h1, h2, postmeans, postsds, nsamp)
  bf <- (fit[1] / comp[1]) / (fit[2] / comp[2])
  return(bf)
}

# Documentation samp_bf() ====
#' Sample multiple datasets and compute the Bayes factor in each
#'
#' @param datasets A number. The number of datasets to simulate for each
#' sample size \code{n}
#' @param n A number. The group sample size to be used in data simulation
#' @param ngroup A number. The number of groups.
#' @param means A vector of expected population means.
#' @param sds A vector of expected population standard deviations
#' Note, when standardized, this is a vector of 1s
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param scale A number specifying the prior scale.
#' @param nsamp A number. The number of samples for the fit and complexity
#' See \code{?BayesianPower::calc_fc}
#' @return A vector of Bayes factors BF12 for each of the simulated datasets
samp_bf <- function(datasets, n, ngroup, means, sds, h1, h2, scale, nsamp) {
  # Function ====
  out <- sapply(1:datasets, function(i){
    data <- matrix(stats::rnorm(n * ngroup, mean = means, sd = sds),
                   ncol = ngroup,
                   byrow = TRUE)
    calc_bf(data = data, h1, h2, scale, nsamp)
  }
  )
  return(out)
}

# Documentation bayes_error() ====
#' Determine the unconditional error probabilities for a set of simulated
#' Bayes factors.
#'
#' @param BFs1 A vector. Simulated BF12 under H1 for a given n
#' @param BFs2 A vector. Simulated BF12 under H2 for a given n
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @return A named vector. The Type 1, Type 2, Decision error and Area of Indecision probabilities
#' and the median Bayes factors under H1 and H2
bayes_error <- function(BFs1, BFs2, bound1 = 1, bound2 = 1/bound1) {
  # Function ====
  type1 <- mean(BFs1 < bound2)
  type2 <- mean(BFs2 > bound1)
  de <- (type1 + type2)/2
  aoi <- if (bound1 != bound2) {
    (mean((BFs1 < bound1) & (BFs1 > bound2)) +
       mean((BFs2 < bound1) & (BFs2 > bound2))) / 2
  } else {
    0
  }
  med.2 <- stats::median(BFs2)
  med.1 <- 1/stats::median(BFs1)
  perc10.1 <- as.numeric(stats::quantile(BFs1, .10))
  perc90.1 <- as.numeric(stats::quantile(BFs1, .90))
  perc10.2 <- as.numeric(stats::quantile(BFs2, .10))
  perc90.2 <- as.numeric(stats::quantile(BFs2, .90))
  return(c("1" = type1, "2" = type2, "de" = de, "aoi" = aoi,
           "med.1" = med.1, "med.2" = med.2,
           "quant10.1" = perc10.1, "quant90.1" = perc90.1,
           "quant10.2" = perc10.2, "quant90.2" = perc90.2))
}


R/background.R
# Documentation eval_const() ====
#' Evaluate a contraint matrix for a set of prior/posterior samples
#'
#' @param hyp A constraint matrix defining a hypothesis.
#' @param samples A matrix. Prior or posterior samples, the number of columns
#' corresponds to the number of groups, the number of rows the number of samples
#' @return A number between 0 and 1. The proportion of samples in which the
#' constraints are met.
eval_const <- function(hyp, samples) {
  # Errors ====
  if(length(dim(hyp)) != 2) {
    stop("Hypothesis must be a matrix")
  }
  if(length(dim(samples)) != 2) {
    stop("Samples must be in a matrix form")
  }
  if(ncol(samples) != ncol(hyp)) {
    stop("Not the right dimensions")
  }
  # Function ====
  nconst <- nrow(hyp)

  out <- mean(apply(samples, 1, function(i) {
    rowCheck <- apply(hyp, 1, function(j) {
      sum(i * j) > 0
    }
    )
    sum(rowCheck) == nconst
  }
  )
  )
  return(out)
}

# Documentation samp_dist() ====
#' Sample from prior or posterior distribution
#'
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param means A vector. The prior or posterior means for each group
#' @param sds A number or a vector. The standard deviations for each group
#' If a number is used, the same prior or posterior standard deviation is
#' used for each group.
#' @return A matrix of \code{nsamp} rows and as many columns as the
#' length of \code{means}. Posterior or prior samples
samp_dist <- function(nsamp, means, sds) {
  # Function ====
  ngroup <- length(means)

  samples <- matrix(rnorm(nsamp * ngroup,
                          mean = means,
                          sd = sds),
                    ncol = ngroup,
                    byrow = TRUE)
  return(samples)
}

# Documentation calc_fc() ====
#' Compute the complexity or fit for two hypotheses.
#' @param hyp A constraint matrix defining H1.
#' @param hyps2 A constraint matrix defining H2 OR a character \code{'u'}
#' or \code{'c'} specifying an unconstrained or complement hypothesis
#' @param means A vector of posterior or prior means
#' @param sds A vector or posterior or prior standard deviation
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @return A vector.
#' The proportion of posterior samples in agreement with H1 and with H2
#' @examples
#' # Complexity
#' priormeans <- c(0,0,0)
#' priorsds <- c(1,1,1)
#' h1 <- matrix(c(1,-1,0,0,1,-1), nrow= 2, byrow= TRUE)
#' h2 <- 'c'
#' complexity <- calc_fc(h1, h2, priormeans, priorsds)
#'
#' # Fit
#' truemeans <- c(.4, .2, 0)
#' n <- 30
#' ngroup <- length(truemeans)
#' dat <- matrix(rnorm(n * ngroup, truemeans), ncol = ngroup, byrow = TRUE)
#' postmeans <- colMeans(dat)
#' postsds <- apply(dat, 2, sd)/sqrt(n)
#' fit <- calc_fc(h1, h2, postmeans, postsds)
#'
calc_fc <- function(hyp, hyp2, means, sds, nsamp = 1000) {
  # Errors ====
  if(is.numeric(hyp2)){
    if(length(dim(hyp2)) != 2) stop("Hypothesis must be a matrix")
  } else {
    if(hyp2 != "c" && hyp2 != "u") stop("Use 'u' or 'c' for unconstrained or complement.")
  }
  if(length(dim(hyp)) != 2) {
    stop("Hypothesis must be a matrix")
  }
  # Function ====
  ngroup <- length(means)
  samples <- samp_dist(nsamp, means = means, sds = sds)
  fc1 <- eval_const(hyp, samples)
  if (fc1 == 0) fc1 <- 1/nsamp
  fc2 <- if (is.character(hyp2)) {
    if (hyp2 == 'u') 1
    if (hyp2 == 'c') 1 - fc1
  } else {
    eval_const(hyp2, samples)
  }
  if (fc2 == 0) fc2 <- 1 / nsamp
  out <- c(fc1, fc2)
  return(out)
}

# Documentation calc_bf()====
#' Compute a Bayes factor
#'
#' @param data A matrix. The dataset for which the BF must be computed
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param comp A vector with the complexity of H1 and H2
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' @return BF12, that is, the evidence for H1 relative to H2
calc_bf <- function(data, h1, h2, comp, nsamp = 1000) {
  # Function ====
  postmeans <- colMeans(data)
  postsds <- apply(data, 1, sd) / sqrt(nrow(data))
  fit <- calc_fc(h1, h2, postmeans, postsds, nsamp)
  bf <- (fit[1] / comp[1]) / (fit[2] / comp[2])
  return(bf)
}

# Documentation samp_bf() ====
#' Sample multiple datasets and compute the Bayes factor in each
#'
#' @param datasets A number. The number of datasets to simulate for each
#' sample size \code{n}
#' @param n A number. The group sample size to be used in data simulation
#' @param ngroup A number. The number of groups.
#' @param means A vector of expected population means (standardized).
#' @param sds A vector of expected population effect sizes.
#' Note, when standardized, this is a vector of 1s
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param comp A vector of the complexities of H1 and H2.
#' See \code{?BayesianPower::calc_fc}
#' @return A vector of Bayes factors BF12 for each of the simulated datasets
samp_bf <- function(datasets, n, ngroup, means, sds, h1, h2, comp) {
  # Function ====
  out <- sapply(1:datasets, function(i){
    data <- matrix(rnorm(n * ngroup, mean = means, sd = 1),
                   ncol = ngroup,
                   byrow = TRUE)
    calc_bf(data = data, h1, h2, comp, nsamp)
  }
  )
  return(out)
}

# Documentation bayes_error() ====
#' Determine the unconditional error probabilities for a set of simulated
#' Bayes factors.
#'
#' @param BFs1 A vector. Simulated BF12 under H1 for a given n
#' @param BFs2 A vector. Simulated BF12 under H2 for a given n
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @return A named vector. The Type 1, Type 2, DE, AOI error probabilities
#' and the median Bayes factors under H1 and H2
bayes_error <- function(BFs1, BFs2, bound1 = 1, bound2 = 1/bound1) {
  # Function ====
  type1 <- mean(BFs1 < bound1)
  type2 <- mean(BFs2 > bound2)
  de <- (type1 + type2)/2
  aoi <- if (bound1 != bound2) {
    (mean((BFs1 > bound1) & (BFs1 < bound2)) +
       mean((BFs2 > bound1) & (BFs2 < bound2))) / 2
  } else {
    0
  }
  med.2.inv <- 1/median(BFs2)
  med.1 <- median(BFs1)
  return(c("1" = type1, "2" = type2, "de" = de, "aoi" = aoi,
           "med.1" = med.1, "med.2 inverse" = med.2.inv))
}


R/samplesize.R
# Bayes_sampsize ====
#' Determine the required sample size for a Bayesian hypothesis test
#'
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param m1 A vector of expected population means under H1 (standardized).
#' @param m2 A vector of expected populations means under H2 (standardized).
#' \code{m2} must be of same length as \code{m1}
#' @param type A character. The type of error to be controlled
#' options are: \code{"1", "2", "de", "aoi", "med.1", "med.2"}
#' @param cutoff A number. The cutoff criterion for type.
#' If \code{type} is \code{"1", "2", "de", "aoi"}, \code{cutoff} must be between 0 and 1
#' If \code{type} is \code{"med.1" or "med.2"}, \code{cutoff} must be larger than 1
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @param datasets A number. The number of datasets to compute the error probabilities
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param minss A number. The minimum sample size to consider
#' @param maxss A number. The maximum sample size to consider
#' @param seed A number. The random seed to be set
#' @return The sample size for which the chosen type of error probability
#' is at the set cutoff, and the according error probabilities and median BFs
#' @examples
#' # Example 1 Decision error and Hc
#' h1 <- matrix(c(1,-1,0,0,1,-1), nrow= 2, byrow= TRUE)
#' h2 <- 'c'
#' m1 <- c(.4,.2,0)
#' m2 <- c(.2,0,.1)
#' bayes_sampsize(h1, h2, m1, m2, type = "de", cutoff = .125, minss = 2, maxss = 1000, datasets = 100)
#'
#' # Example 2 Indecision error and H2
#' h1 <- matrix(c(1,-1,0,0,0,1,-1,0,0,0,1,-1), nrow= 3, byrow= TRUE)
#' h2 <- matrix(c(0,-1,1,0,0,1,0,-1,-1,0,0,1), nrow = 3, byrow= TRUE)
#' m1 <- c(.7,.3,.1,0)
#' m2 <- c(0,.4,.5,.1)
#' bayes_sampsize(h1, h2, m1, m2, type = "aoi", cutoff = .2, minss = 2, maxss = 500, datasets = 1000)
#' @export
bayes_sampsize <- function(h1, h2, m1, m2,
                        type = 1, cutoff, bound1 = 1, bound2 = 1 / bound1,
                        datasets = 1000, nsamp = 1000,
                        minss = 2, maxss = 1000, seed = 31) {
  # Errors ====
  if(!is.numeric(cutoff)) stop("expected numeric value")
  if(!is.numeric(minss) || !is.numeric(maxss)) stop("minss and maxss must be integer")

  types <- c("1", "2", "de", "aoi", "med.1", "med.2")
  if(all(types != type)) stop("Incorrect type specified")

  if(type == "med.1" || type == "med.2") {
    if(cutoff < 1) stop("Cutoff must be larger than 1 for controlling median BF")
  } else {
    if(cutoff > 1) stop( "Cutoff must be smaller than 1 for controlling error probabilities")
  }

  if((maxss-minss) < 20 ) stop("Difference between minss and maxss must be at least 20")

  # Function ====
  times <- c(ceiling(log(maxss) / log(minss + 1)) + 1)
  timesT <- 0
  {
    pb <- utils::txtProgressBar(min = 0, max = times)

    set.seed(seed)
    n <- round(maxss / 2)
    oldn <- n + 10

    ngroup = length(m1)
    priormeans <- rep(0, ngroup)
    priorsd <- rep(1000, ngroup)

    comp <- calc_fc(h1, h2, means = priormeans, sds = priorsd, nsamp = nsamp)

    while ((abs(oldn - n) > 1)) {
      errors <- bayes_power(n = n, h1 = h1, h2 = h2, m1 = m1, m2 = m2,
                            ngroup = ngroup, comp = comp,
                            bound1 = bound1, bound2 = bound2,
                            datasets = datasets, nsamp = nsamp)
      if (errors[[paste(type)]] < cutoff) {
        oldn <- n
        maxss <- n
        n <- round((n + minss) / 2)
      }
      if (errors[[paste(type)]] > cutoff) {
        oldn <- n
        minss <- n
        n <- round((n + maxss) / 2)
      }
      if (errors[[paste(type)]] == cutoff) break

      timesT <- timesT + 1
        utils::setTxtProgressBar(pb, timesT)
    }
    close(pb)
  }
  return(c("Sample size" = as.integer(n), "Type 1 error" = errors[[1]],
           "Type 2 error" = errors[[2]], "Decision error" = errors[[3]],
           "Indecision error" = errors[[4]],
           "Median BF12 under H1" = errors[[5]],
           "Median BF21 under H2"= 1 / errors[[6]]))
}

# Bayes_power ====
#' Determine the 'power' for a Bayesian hypothesis test
#'
#' @param n A number. The sample size
#' @param h1 A constraint matrix defining H1
#' @param h2 A constraint matrix defining H2
#' @param m1 A vector of expected population means under H1 (standardized)
#' @param m2 A vector of expected populations means under H2 (standardized)
#' \code{m2} must be of same length as \code{m1}
#' @param ngroup A number or \code{NULL}. The number of groups
#' If \code{NULL} the number of groups is determined from the length of \code{m1}
#' @param comp A vector or \code{NULL}. The complexity of H1 and H2. If
#' \code{NULL} the complexity is also estimated
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @param datasets A number. The number of datasets to compute the error probabilities
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param seed A number. The random seed to be set
#' @return The Type 1, Type 2, DE and AOI error probability and
#' the median BF12s under H1 and H2
#' @examples
#' # Example 1 H1 vs Hc
#' h1 <- matrix(c(1,-1,0,0,1,-1), nrow= 2, byrow= TRUE)
#' h2 <- "c"
#' m1 <- c(.4,.2,0)
#' m2 <- c(.2,0,.1)
#' bayes_power(40, h1, h2, m1, m2)
#'
#' # Example 2 H1 vs H2
#' h1 <- matrix(c(1,-1,0,0,0,1,-1,0,0,0,1,-1), nrow= 3, byrow= TRUE)
#' h2 <- matrix(c(0,-1,1,0,0,1,0,-1,-1,0,0,1), nrow = 3, byrow= TRUE)
#' m1 <- c(.7,.3,.1,0)
#' m2 <- c(0,.4,.5,.1)
#' bayes_power(40, h1, h2, m1, m2)
#' @export
bayes_power <- function(n, h1, h2, m1, m2,
                        ngroup = NULL, comp = NULL,
                        bound1 = 1, bound2 = 1/bound1,
                        datasets = 1000, nsamp = 1000,
                        seed = NULL){

  # Errors ====
  if(!is.numeric(bound1) || !is.numeric(bound2) ||
     !is.numeric(m1) || !is.numeric(m2) || !is.numeric(h1) ||
     !is.numeric(datasets) || !is.numeric(n) ||
     !is.numeric(nsamp)) {
    stop("expected numeric value")
  }
    if(length(dim(h1)) != 2) stop("h1 must be a matrix")

  if(is.numeric(h2)){
    if(length(dim(h2)) != 2) stop("h2 must be a matrix")
  } else {
    if(h2 != "c" && h2 != "u") stop("Use 'u' or 'c' for unconstrained or complement.")
  }

  if(length(m1) != length(m2)) stop("m1 and m2 must be the same length")
  if(length(m1) != ncol(h1)) stop("h1 and m1 do not match")

  if(bound1 < bound2) stop("bound1 must be larger than bound2")

   # Function ====
  if(!is.null(seed)) {
    if(!is.integer(seed)) stop("seed must be integer")
    set.seed(seed)
  }
  if(is.null(ngroup)) {
    ngroup <- length(m1)
  } else {
    if(!is.integer(ngroup)) stop("ngroup must be integer")
  }
  if(is.null(comp)) {
    priormeans <- rep(0, ngroup)
    priorsd <- rep(1000, ngroup)
    comp <- calc_fc(h1, h2, means = priormeans, sds = priorsd, nsamp = nsamp)
  } else {
    if(!is.numeric(comp)) stop("comp must be numeric")
  }

  BF1 <- samp_bf(datasets, n, ngroup, means = m1, sds = 1, h1, h2, comp, nsamp)
  BF2 <- samp_bf(datasets, n, ngroup, means = m2, sds = 1, h1, h2, comp, nsamp)
  errors <- bayes_error(BF1, BF2, bound1, bound2)
  return(errors)
}


R/power_sampsize.R
# Bayes_sampsize ====
#' Determine the required sample size for a Bayesian hypothesis test
#'
#' @param h1 A constraint matrix defining H1.
#' @param h2 A constraint matrix defining H2.
#' @param m1 A vector of expected population means under H1 (standardized).
#' @param m2 A vector of expected populations means under H2 (standardized).
#' \code{m2} must be of same length as \code{m1}
#' @param sd1 A vector of standard deviations under H1. Must be a single number (equal
#' standard deviation under all populations), or a vector of the same length as \code{m1}
#' @param sd2 A vector of standard deviations under H2. Must be a single number (equal
#' standard deviation under all populations), or a vector of the same length as \code{m2}
#' @param scale A number specifying the prior scale
#' @param type A character. The type of error to be controlled
#' options are: \code{"1", "2", "de", "aoi", "med.1", "med.2"}
#' @param cutoff A number. The cutoff criterion for type.
#' If \code{type} is \code{"1", "2", "de", "aoi"}, \code{cutoff} must be between 0 and 1
#' If \code{type} is \code{"med.1" or "med.2"}, \code{cutoff} must be larger than 1
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @param datasets A number. The number of datasets to compute the error probabilities
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param minss A number. The minimum sample size to consider
#' @param maxss A number. The maximum sample size to consider
#' @param seed A number. The random seed to be set
#' @return The sample size for which the chosen type of error probability
#' is at the set cutoff, and the according error probabilities and median Bayes factors
#' @examples
#' # Short computation example NOT SUFFICIENT SAMPLES
#' h1 <- matrix(c(1,-1), nrow= 1, byrow= TRUE)
#' h2 <- 'c'
#' m1 <- c(.4, 0)
#' m2 <- c(0, .1)
#' bayes_sampsize(h1, h2, m1, m2, sd1 = 1, sd2 = 1, scale = 1000,
#' type = "de", cutoff = .125, nsamp = 50, datasets = 50,
#' minss = 40, maxss = 70)
#' @export
bayes_sampsize <- function(h1, h2, m1, m2, sd1 = 1, sd2 = 1, scale = 1000,
                        type = 1, cutoff, bound1 = 1, bound2 = 1 / bound1,
                        datasets = 1000, nsamp = 1000,
                        minss = 2, maxss = 1000, seed = 31) {
  # Errors ====
  if(!is.numeric(cutoff)) stop("expected numeric value")
  if(!is.numeric(minss) || !is.numeric(maxss)) stop("minss and maxss must be integer")

  types <- c("1", "2", "de", "aoi", "med.1", "med.2")
  if(all(types != type)) stop("Incorrect type specified")

  if(type == "med.1" || type == "med.2") {
    if(cutoff < 1) stop("Cutoff must be larger than 1 for controlling median BF")
    cutoff <- 1/cutoff
  } else {
    if(cutoff > 1) stop( "Cutoff must be smaller than 1 for controlling error probabilities")
  }

  if((maxss-minss) < 20 ) stop("Difference between minss and maxss must be at least 20")

  # Function ====
  times <- c(ceiling(log(maxss) / log(minss + 1)) + 1)
  timesT <- 0
  {
    pb <- utils::txtProgressBar(min = 0, max = times)

    set.seed(seed)
    n <- round(maxss / 2)
    oldn <- n + 10

    ngroup = length(m1)

    while ((abs(oldn - n) > 1)) {
      errors <- bayes_power(n = n, h1 = h1, h2 = h2, m1 = m1, m2 = m2,
                            sd1 = sd1, sd2 = sd2, scale = scale,
                            bound1 = bound1, bound2 = bound2,
                            datasets = datasets, nsamp = nsamp, seed = seed)
      if (errors[[paste(type)]] < cutoff) {
        oldn <- n
        maxss <- n
        n <- round((n + minss) / 2)
      }
      if (errors[[paste(type)]] > cutoff) {
        oldn <- n
        minss <- n
        n <- round((n + maxss) / 2)
      }
      if (errors[[paste(type)]] == cutoff) break

      timesT <- timesT + 1
        utils::setTxtProgressBar(pb, timesT)
    }
    close(pb)
  }
  return(c("Sample size" = as.integer(n), "Type 1 error" = errors[[1]],
           "Type 2 error" = errors[[2]], "Decision error" = errors[[3]],
           "Indecision error" = errors[[4]],
           "Median BF12 under H1" = errors[[5]],
           "Median BF21 under H2"= 1 / errors[[6]],
         "10quantile under H1" = errors[[7]],
         "90quantile under H1" = errors[[8]],
         "10quantile under H2" = errors[[9]],
         "90quantile under H2" = errors[[10]]))
}

# Bayes_power ====
#' Determine the 'power' for a Bayesian hypothesis test
#'
#' @param n A number. The sample size
#' @param h1 A constraint matrix defining H1
#' @param h2 A constraint matrix defining H2
#' @param m1 A vector of expected population means under H1
#' @param m2 A vector of expected populations means under H2
#' \code{m2} must be of same length as \code{m1}
#' @param sd1 A vector of standard deviations under H1. Must be a single number (equal
#' standard deviation under all populations), or a vector of the same length as \code{m1}
#' @param sd2 A vector of standard deviations under H2. Must be a single number (equal
#' standard deviation under all populations), or a vector of the same length as \code{m2}
#' @param scale A number specifying the prior scale
#' @param bound1 A number. The boundary above which BF12 favors H1
#' @param bound2 A number. The boundary below which BF12 favors H2
#' @param datasets A number. The number of datasets to compute the error probabilities
#' @param nsamp A number. The number of prior or posterior samples to determine the
#' fit and complexity
#' @param seed A number. The random seed to be set
#' @return The Type 1, Type 2, Decision error and Area of Indecision probability and
#' the median BF12s under H1 and H2
#' @examples
#' # Short example WITH SMALL AMOUNT OF SAMPLES
#' h1 <- matrix(c(1,-1,0,0,1,-1), nrow= 2, byrow= TRUE)
#' h2 <- "c"
#' m1 <- c(.4,.2,0)
#' m2 <- c(.2,0,.1)
#' bayes_power(40, h1, h2, m1, m2, datasets = 50, nsamp = 50)
#' @export
bayes_power <- function(n, h1, h2, m1, m2,
                        sd1 = 1, sd2 = 1, scale = 1000,
                        bound1 = 1, bound2 = 1/bound1,
                        datasets = 1000, nsamp = 1000,
                        seed = 31){

  # Errors ====
  if(!is.numeric(bound1) || !is.numeric(bound2) ||
     !is.numeric(m1) || !is.numeric(m2) || !is.numeric(h1) ||
     !is.numeric(datasets) || !is.numeric(n) ||
     !is.numeric(nsamp) || !is.numeric(sd1) || !is.numeric(sd2) ||
     !is.numeric(scale) || !is.numeric(seed)) {
    stop("expected numeric value")
  }
  if(round(n) != n) stop("n must be integer")
  if(length(dim(h1)) != 2) stop("h1 must be a matrix")
  if(any(round(h1) != h1)) stop("h1 can only contain integers")
  if(any(h1 > 1) || any(h1 < -1)) stop( "h1 must consist of 1 -1 and 0")

  if(is.numeric(h2)){
    if(length(dim(h2)) != 2) stop("h2 must be a matrix")
    if(any(round(h2) != h2)) stop("h2 can only contain integers")
    if(any(h2 > 1) || any(h2 < -1)) stop( "h2 must consist of 1 -1 and 0")
  } else {
    if(h2 != "c" && h2 != "u") stop("Use 'u' or 'c' for unconstrained or complement.")
  }

  if(length(m1) != length(m2)) stop("m1 and m2 must be the same length")
  if(length(m1) != ncol(h1)) stop("h1 and m1 do not match")

  if(length(sd1) != 1){
    if(length(sd1 != length(m1))) stop("sd1 must be a single number or a vector of same length as m1")
  }
  if(length(sd2) != 1){
    if(length(sd2 != length(m2))) stop("sd2 must be a single number or a vector of same length as m2")
  }

  if(bound1 < bound2) stop("bound1 must be larger than bound2")

  #  # Function ====
  # if(!is.null(seed)) {
  #   if(!is.integer(seed)) stop("seed must be integer")
     set.seed(seed)
  # }
  # if(is.null(ngroup)) {
     ngroup <- length(m1)
  # } else {
  #   if(!is.integer(ngroup)) stop("ngroup must be integer")
  # }

  BF1 <- samp_bf(datasets, n, ngroup, means = m1, sds = sd1, h1, h2, scale, nsamp)
  BF2 <- samp_bf(datasets, n, ngroup, means = m2, sds = sd2, h1, h2, scale, nsamp)
  errors <- bayes_error(BF1, BF2, bound1, bound2)
  return(errors)
}
